src/stream.cpp

```diff
@@ -127,11 +127,50 @@ namespace oxen::quic
         });
     }
 
+    void Stream::stop_reading()
+    {
+        endpoint.call([this]() {
+            if (not _is_reading)
+            {
+                log::warning(log_cat, "Stream has already halted read operations!");
+                return;
+            }
+
+            _is_reading = false;
+
+            log::warning(log_cat, "Halting all read operations on stream ID:{}!", _stream_id);
+            ngtcp2_conn_shutdown_stream_read(*_conn, 0, _stream_id, 0);
+        });
+    }
+
+    void Stream::stop_writing()
+    {
+        endpoint.call([this]() {
+            if (not _is_writing)
+            {
+                log::warning(log_cat, "Stream has already halted write operations!");
+                return;
+            }
+
+            _is_writing = false;
```

Comment:
This needs to check whether the "can we shutdown write" condition (`user_buffers.empty()` — see other comment) is already satisfied and, if so, initiate the shutdown immediately.  Otherwise, if you shutdown after everything you sent has already been acked, there will never be another ack to trigger the actual stream shutdown code.